# 2014222 - Semester 2 CA-01 -April 2024¶

# Part 2

# Step 1: Import Python Libraries

In [1]:
# Utilise Python programming language in order to comply with the requisites of the assessment and perform adequate Machine
# Learning algorithms to discover and deliver insights.

# Import the necessary libraries (Pandas) in order to perform data cleansing. These are the libraries that are 
# conventionally used as a common practice in order to perform mathematical and statistical operations during a data analysis
# process

import pandas as pd

# Import the Plotly library in order to perform data visualisation procedures

import plotly.express as px
import plotly.graph_objects as go

# I am using this line of code to see all columns in a wide DataFrame

pd.set_option('display.max_columns', None)

# Step 2: Reading Dataset

In [2]:
# Read the CSV file by applying the pd.read_csv() function.

nba_data = pd.read_csv('NBA Dataset CSV C1.csv')

Reference to the dataset:

https://stats.nba.com/stats/leagueLeaders?LeagueID=00&PerMode=Totals&Scope=S&Season=2012-13&SeasonType=Regular%20Season&StatCategory=PTS

# Analyzing the Data

In [3]:
# Print the NBA table dataframe

nba_data

Year       Season_type  PLAYER_ID  RANK           PLAYER     TEAM_ID  \
0     2012-13  Regular%20Season     201142     1     Kevin Durant  1610612760   
1     2012-13  Regular%20Season        977     2      Kobe Bryant  1610612747   
2     2012-13  Regular%20Season       2544     3     LeBron James  1610612748   
3     2012-13  Regular%20Season     201935     4     James Harden  1610612745   
4     2012-13  Regular%20Season       2546     5  Carmelo Anthony  1610612752   
...       ...               ...        ...   ...              ...         ...   
7288  2021-22          Playoffs    1629006   206      Josh Okogie  1610612750   
7289  2021-22          Playoffs    1630556   206  Kessler Edwards  1610612751   
7290  2021-22          Playoffs    1630201   206    Malachi Flynn  1610612761   
7291  2021-22          Playoffs     202693   206  Markieff Morris  1610612748   
7292  2021-22          Playoffs     200794   206     Paul Millsap  1610612755   

     TEAM  GP   MIN  FGM   FGA  FG_PCT  FG3M  FG3A  FG3_PCT  FTM  FTA  FT_PCT  \
0     OKC  81  3119  731  1433   0.510   139   334    0.416  679  750   0.905   
1     LAL  78  3013  738  1595   0.463   132   407    0.324  525  626   0.839   
2     MIA  76  2877  765  1354   0.565   103   254    0.406  403  535   0.753   
3     HOU  78  2985  585  1337   0.438   179   486    0.368  674  792   0.851   
4     NYK  67  2482  669  1489   0.449   157   414    0.379  425  512   0.830   
...   ...  ..   ...  ...   ...     ...   ...   ...      ...  ...  ...     ...   
7288  MIN   1     2    0     0   0.000     0     0    0.000    0    0   0.000   
7289  BKN   2     7    0     0   0.000     0     0    0.000    0    0   0.000   
7290  TOR   6    36    0     7   0.000     0     3    0.000    0    0   0.000   
7291  MIA   2     3    0     1   0.000     0     0    0.000    0    0   0.000   
7292  PHI   1     6    0     0   0.000     0     0    0.000    0    0   0.000   

      OREB  DREB  REB  AST  STL  BLK  TOV   PF   PTS   EFF  AST_TOV  STL_TOV  
0       46   594  640  374  116  105  280  143  2280  2462     1.34     0.41  
1       66   367  433  469  106   25  287  173  2133  1921     1.63     0.37  
2       97   513  610  551  129   67  226  110  2036  2446     2.44     0.57  
3       62   317  379  455  142   38  295  178  2023  1872     1.54     0.48  
4      134   326  460  171   52   32  175  205  1920  1553     0.98     0.30  
...    ...   ...  ...  ...  ...  ...  ...  ...   ...   ...      ...      ...  
7288     0     0    0    0    0    0    0    0     0     0     0.00     0.00  
7289     0     0    0    1    1    0    1    3     0     1     1.00     1.00  
7290     2     1    3    3    1    0    1    6     0    -1     3.00     1.00  
7291     0     1    1    0    0    0    1    2     0    -1     0.00     0.00  
7292     0     1    1    1    0    0    0    1     0     2     0.00     0.00  

[7293 rows x 30 columns]

In [4]:
# Print the dimensions of the NBA DataFrame.

nba_data.shape

(7293, 30)

As we can see that the NBA dataframe contains 7293 rows × 30 columns

In [5]:
# Print the first 5 rows of the NBA dataframe by applying the .head() method,
# This method is will display the top 5 observations of the dataset

nba_data.head()

Year       Season_type  PLAYER_ID  RANK           PLAYER     TEAM_ID  \
0  2012-13  Regular%20Season     201142     1     Kevin Durant  1610612760   
1  2012-13  Regular%20Season        977     2      Kobe Bryant  1610612747   
2  2012-13  Regular%20Season       2544     3     LeBron James  1610612748   
3  2012-13  Regular%20Season     201935     4     James Harden  1610612745   
4  2012-13  Regular%20Season       2546     5  Carmelo Anthony  1610612752   

  TEAM  GP   MIN  FGM   FGA  FG_PCT  FG3M  FG3A  FG3_PCT  FTM  FTA  FT_PCT  \
0  OKC  81  3119  731  1433   0.510   139   334    0.416  679  750   0.905   
1  LAL  78  3013  738  1595   0.463   132   407    0.324  525  626   0.839   
2  MIA  76  2877  765  1354   0.565   103   254    0.406  403  535   0.753   
3  HOU  78  2985  585  1337   0.438   179   486    0.368  674  792   0.851   
4  NYK  67  2482  669  1489   0.449   157   414    0.379  425  512   0.830   

   OREB  DREB  REB  AST  STL  BLK  TOV   PF   PTS   EFF  AST_TOV  STL_TOV  
0    46   594  640  374  116  105  280  143  2280  2462     1.34     0.41  
1    66   367  433  469  106   25  287  173  2133  1921     1.63     0.37  
2    97   513  610  551  129   67  226  110  2036  2446     2.44     0.57  
3    62   317  379  455  142   38  295  178  2023  1872     1.54     0.48  
4   134   326  460  171   52   32  175  205  1920  1553     0.98     0.30

In [6]:
# Print the last 5 rows of the NBA dataframe by applying the .tail() method,
# Will display the last 5 observations of the dataset

nba_data.tail()

Year Season_type  PLAYER_ID  RANK           PLAYER     TEAM_ID TEAM  \
7288  2021-22    Playoffs    1629006   206      Josh Okogie  1610612750  MIN   
7289  2021-22    Playoffs    1630556   206  Kessler Edwards  1610612751  BKN   
7290  2021-22    Playoffs    1630201   206    Malachi Flynn  1610612761  TOR   
7291  2021-22    Playoffs     202693   206  Markieff Morris  1610612748  MIA   
7292  2021-22    Playoffs     200794   206     Paul Millsap  1610612755  PHI   

      GP  MIN  FGM  FGA  FG_PCT  FG3M  FG3A  FG3_PCT  FTM  FTA  FT_PCT  OREB  \
7288   1    2    0    0     0.0     0     0      0.0    0    0     0.0     0   
7289   2    7    0    0     0.0     0     0      0.0    0    0     0.0     0   
7290   6   36    0    7     0.0     0     3      0.0    0    0     0.0     2   
7291   2    3    0    1     0.0     0     0      0.0    0    0     0.0     0   
7292   1    6    0    0     0.0     0     0      0.0    0    0     0.0     0   

      DREB  REB  AST  STL  BLK  TOV  PF  PTS  EFF  AST_TOV  STL_TOV  
7288     0    0    0    0    0    0   0    0    0      0.0      0.0  
7289     0    0    1    1    0    1   3    0    1      1.0      1.0  
7290     1    3    3    1    0    1   6    0   -1      3.0      1.0  
7291     1    1    0    0    0    1   2    0   -1      0.0      0.0  
7292     1    1    1    0    0    0   1    0    2      0.0      0.0

In [7]:
# Getting information about the NBA dataframe by applying the .info() method,
# Will display number of records in each column, data having null or not null, Data type,
# memory usage of the dataset
# helps to understand the data type and information about data

nba_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7293 entries, 0 to 7292
Data columns (total 30 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Year         7293 non-null   object 
 1   Season_type  7293 non-null   object 
 2   PLAYER_ID    7293 non-null   int64  
 3   RANK         7293 non-null   int64  
 4   PLAYER       7293 non-null   object 
 5   TEAM_ID      7293 non-null   int64  
 6   TEAM         7293 non-null   object 
 7   GP           7293 non-null   int64  
 8   MIN          7293 non-null   int64  
 9   FGM          7293 non-null   int64  
 10  FGA          7293 non-null   int64  
 11  FG_PCT       7293 non-null   float64
 12  FG3M         7293 non-null   int64  
 13  FG3A         7293 non-null   int64  
 14  FG3_PCT      7293 non-null   float64
 15  FTM          7293 non-null   int64  
 16  FTA          7293 non-null   int64  
 17  FT_PCT       7293 non-null   float64
 18  OREB         7293 non-null   int64  
 19  DREB  

Seems to be no missing values.

This information can be helpful for diagnosing issues with data processing and understanding the structure of the NBA dataset.


# Check for Duplication

In [8]:
# Getting several unique values in each columnan the data description the NBA dataframe
# by applying the .nunique() method,
# Will display continuous and categorical columns in the data.
# Duplicated data can be handled or removed based on further analysis
# helps to understand the data type and information about data

nba_data.nunique()

Year             10
Season_type       2
PLAYER_ID      1386
RANK            537
PLAYER         1382
TEAM_ID          30
TEAM             31
GP               83
MIN            2331
FGM             670
FGA            1210
FG_PCT          483
FG3M            261
FG3A            566
FG3_PCT         379
FTM             428
FTA             517
FT_PCT          539
OREB            285
DREB            590
REB             748
AST             544
STL             165
BLK             175
TOV             289
PF              276
PTS            1449
EFF            1567
AST_TOV         445
STL_TOV         226
dtype: int64

In [9]:
# Identifying null values in the NBA dataframe by applying the .isnull() method,

nba_data.isnull()

Year  Season_type  PLAYER_ID   RANK  PLAYER  TEAM_ID   TEAM     GP  \
0     False        False      False  False   False    False  False  False   
1     False        False      False  False   False    False  False  False   
2     False        False      False  False   False    False  False  False   
3     False        False      False  False   False    False  False  False   
4     False        False      False  False   False    False  False  False   
...     ...          ...        ...    ...     ...      ...    ...    ...   
7288  False        False      False  False   False    False  False  False   
7289  False        False      False  False   False    False  False  False   
7290  False        False      False  False   False    False  False  False   
7291  False        False      False  False   False    False  False  False   
7292  False        False      False  False   False    False  False  False   

        MIN    FGM    FGA  FG_PCT   FG3M   FG3A  FG3_PCT    FTM    FTA  \
0     False  False  False   False  False  False    False  False  False   
1     False  False  False   False  False  False    False  False  False   
2     False  False  False   False  False  False    False  False  False   
3     False  False  False   False  False  False    False  False  False   
4     False  False  False   False  False  False    False  False  False   
...     ...    ...    ...     ...    ...    ...      ...    ...    ...   
7288  False  False  False   False  False  False    False  False  False   
7289  False  False  False   False  False  False    False  False  False   
7290  False  False  False   False  False  False    False  False  False   
7291  False  False  False   False  False  False    False  False  False   
7292  False  False  False   False  False  False    False  False  False   

      FT_PCT   OREB   DREB    REB    AST    STL    BLK    TOV     PF    PTS  \
0      False  False  False  False  False  False  False  False  False  False   
1      False  False  False  False  False  False  False  False  False  False   
2      False  False  False  False  False  False  False  False  False  False   
3      False  False  False  False  False  False  False  False  False  False   
4      False  False  False  False  False  False  False  False  False  False   
...      ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
7288   False  False  False  False  False  False  False  False  False  False   
7289   False  False  False  False  False  False  False  False  False  False   
7290   False  False  False  False  False  False  False  False  False  False   
7291   False  False  False  False  False  False  False  False  False  False   
7292   False  False  False  False  False  False  False  False  False  False   

        EFF  AST_TOV  STL_TOV  
0     False    False    False  
1     False    False    False  
2     False    False    False  
3     False    False    False  
4     False    False    False  
...     ...      ...      ...  
7288  False    False    False  
7289  False    False    False  
7290  False    False    False  
7291  False    False    False  
7292  False    False    False  

[7293 rows x 30 columns]

As we can see the NBA dataframe null values are false, meaning there is not missing values

In [10]:
# Identifying missing values in the NBA dataframe by applying the .isna().sum() methods,
# I am using this to get the number of missing records in each column

nba_data.isnull().sum()

Year           0
Season_type    0
PLAYER_ID      0
RANK           0
PLAYER         0
TEAM_ID        0
TEAM           0
GP             0
MIN            0
FGM            0
FGA            0
FG_PCT         0
FG3M           0
FG3A           0
FG3_PCT        0
FTM            0
FTA            0
FT_PCT         0
OREB           0
DREB           0
REB            0
AST            0
STL            0
BLK            0
TOV            0
PF             0
PTS            0
EFF            0
AST_TOV        0
STL_TOV        0
dtype: int64

As we can see the NBA dataframe does not contain any duplicates.

In [11]:
# Identifying the percentage of missing values in the NBA dataframe by applying the .isna().sum() methods,
# I am using this to get the number of missing records in each column

(nba_data.isnull().sum()/(len(nba_data)))* 100

Year           0.0
Season_type    0.0
PLAYER_ID      0.0
RANK           0.0
PLAYER         0.0
TEAM_ID        0.0
TEAM           0.0
GP             0.0
MIN            0.0
FGM            0.0
FGA            0.0
FG_PCT         0.0
FG3M           0.0
FG3A           0.0
FG3_PCT        0.0
FTM            0.0
FTA            0.0
FT_PCT         0.0
OREB           0.0
DREB           0.0
REB            0.0
AST            0.0
STL            0.0
BLK            0.0
TOV            0.0
PF             0.0
PTS            0.0
EFF            0.0
AST_TOV        0.0
STL_TOV        0.0
dtype: float64

As we can see the percentage of missing values for the columns is 0%

In [12]:
# Checking if there any missing or null values in the NBA dataframe by applying the .isna() method,

# and the .sum() method to return the sum of the duplicate values in the whole dataframe.

# The duplicated() method would return a series of 'True' or 'False', based on whether a row contains
# a duplicate value or not. Whereas, by adding the .sum() method, which
# is conventionally used to obtain the sum of a series of values,
# it is possible to obtain the whole sum of all the duplicated
# values contained in the dataframe

nba_data.isna().sum()

Year           0
Season_type    0
PLAYER_ID      0
RANK           0
PLAYER         0
TEAM_ID        0
TEAM           0
GP             0
MIN            0
FGM            0
FGA            0
FG_PCT         0
FG3M           0
FG3A           0
FG3_PCT        0
FTM            0
FTA            0
FT_PCT         0
OREB           0
DREB           0
REB            0
AST            0
STL            0
BLK            0
TOV            0
PF             0
PTS            0
EFF            0
AST_TOV        0
STL_TOV        0
dtype: int64

# Step 3: Data Reduction

In [13]:
# Print 10 random records of the NBA table dataframe, by applying the .sample() method
# I want to know which columns or variables can be dropped if they are not useful.

nba_data.sample(10)

Year       Season_type  PLAYER_ID  RANK              PLAYER  \
7086  2021-22          Playoffs     203935    11        Marcus Smart   
92    2012-13  Regular%20Season     203079    93        Dion Waiters   
747   2013-14  Regular%20Season     201150    71       Spencer Hawes   
3498  2017-18  Regular%20Season     202326    45    DeMarcus Cousins   
2129  2015-16  Regular%20Season     201568    68    Danilo Gallinari   
2939  2016-17  Regular%20Season     200752   187            Rudy Gay   
695   2013-14  Regular%20Season     202681    19        Kyrie Irving   
4158  2017-18          Playoffs     203092   163        Tyler Zeller   
5402  2019-20  Regular%20Season    1626187   456  Michael Frazier II   
2109  2015-16  Regular%20Season     203918    48         Rodney Hood   

         TEAM_ID TEAM  GP   MIN  FGM   FGA  FG_PCT  FG3M  FG3A  FG3_PCT  FTM  \
7086  1610612738  BOS  21   760  113   279   0.405    48   137    0.350   50   
92    1610612739  CLE  61  1756  336   815   0.412    63   203    0.310  159   
747   1610612739  CLE  80  2470  403   883   0.456   128   308    0.416  119   
3498  1610612740  NOP  48  1737  406   864   0.470   104   294    0.354  294   
2129  1610612743  DEN  53  1839  287   700   0.410    87   239    0.364  375   
2939  1610612758  SAC  30  1013  201   442   0.455    42   113    0.372  118   
695   1610612739  CLE  71  2496  532  1237   0.430   123   344    0.358  291   
4158  1610612749  MIL   7    66    4     5   0.800     0     0    0.000    3   
5402  1610612745  HOU  13   145    7    29   0.241     4    23    0.174    9   
2109  1610612762  UTA  79  2541  408   972   0.420   161   449    0.359  172   

      FTA  FT_PCT  OREB  DREB  REB  AST  STL  BLK  TOV   PF   PTS   EFF  \
7086   62   0.806    20    75   95  123   26    5   52   71   324   343   
92    213   0.746    22   127  149  183   59   16  121  113   894   647   
747   152   0.783   131   529  660  240   44   94  173  237  1053  1405   
3498  394   0.746   105   512  617  257   79   76  242  183  1210  1439   
2129  432   0.868    52   229  281  135   42   19   77   87  1036   966   
2939  138   0.855    35   155  190   83   44   26   74   79   562   570   
695   338   0.861    52   207  259  433  108   23  190  163  1478  1359   
4158    4   0.750     2    12   14    3    4    3    1   11    11    32   
5402   14   0.643     1     9   10    2    2    0    5   22    27     9   
2109  200   0.860    38   229  267  213   73   16  125  202  1149  1001   

      AST_TOV  STL_TOV  
7086     2.37     0.50  
92       1.51     0.49  
747      1.39     0.25  
3498     1.06     0.33  
2129     1.75     0.55  
2939     1.12     0.59  
695      2.28     0.57  
4158     3.00     4.00  
5402     0.40     0.40  
2109     1.70     0.58

In [14]:
# Droppin the columns that I am not going to use in the NBA dataframe by applying the .drop() method,

# In this case is raking and efficiency columns

nba_data.drop(columns=['RANK','EFF'], inplace=True)

In [15]:
# Getting information about the NBA dataframe by applying the .info() method,
# Will display number of records in each column, data having null or not null, Data type,
# memory usage of the dataset
# helps to understand the data type and information about data

# I am applying this to check is the .drop() method is working

nba_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7293 entries, 0 to 7292
Data columns (total 28 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Year         7293 non-null   object 
 1   Season_type  7293 non-null   object 
 2   PLAYER_ID    7293 non-null   int64  
 3   PLAYER       7293 non-null   object 
 4   TEAM_ID      7293 non-null   int64  
 5   TEAM         7293 non-null   object 
 6   GP           7293 non-null   int64  
 7   MIN          7293 non-null   int64  
 8   FGM          7293 non-null   int64  
 9   FGA          7293 non-null   int64  
 10  FG_PCT       7293 non-null   float64
 11  FG3M         7293 non-null   int64  
 12  FG3A         7293 non-null   int64  
 13  FG3_PCT      7293 non-null   float64
 14  FTM          7293 non-null   int64  
 15  FTA          7293 non-null   int64  
 16  FT_PCT       7293 non-null   float64
 17  OREB         7293 non-null   int64  
 18  DREB         7293 non-null   int64  
 19  REB   

As you can see the columns **'RANK','EFF'** were sucessfully dropped sucess

Resulting in a current total of **28 columns**

In [16]:
# Converts the column called 'Year' from object to integer by applying the 'astype()' method
# and pass 'int' to represent integers.

nba_data['Year'].str[:4].astype(int)

0       2012
1       2012
2       2012
3       2012
4       2012
        ... 
7288    2021
7289    2021
7290    2021
7291    2021
7292    2021
Name: Year, Length: 7293, dtype: int64

The dataype inside the column called 'Year'  now represents **integers.**

In [17]:
# Now that we converted year to represent integers
# I can create a new column called 'season_start_year'

nba_data['season_start_year'] = nba_data['Year'].str[:4].astype(int)

In [18]:
# Checking if the column season_start_year was successfully created
# by applying the .head() method,

nba_data.head()

Year       Season_type  PLAYER_ID           PLAYER     TEAM_ID TEAM  GP  \
0  2012-13  Regular%20Season     201142     Kevin Durant  1610612760  OKC  81   
1  2012-13  Regular%20Season        977      Kobe Bryant  1610612747  LAL  78   
2  2012-13  Regular%20Season       2544     LeBron James  1610612748  MIA  76   
3  2012-13  Regular%20Season     201935     James Harden  1610612745  HOU  78   
4  2012-13  Regular%20Season       2546  Carmelo Anthony  1610612752  NYK  67   

    MIN  FGM   FGA  FG_PCT  FG3M  FG3A  FG3_PCT  FTM  FTA  FT_PCT  OREB  DREB  \
0  3119  731  1433   0.510   139   334    0.416  679  750   0.905    46   594   
1  3013  738  1595   0.463   132   407    0.324  525  626   0.839    66   367   
2  2877  765  1354   0.565   103   254    0.406  403  535   0.753    97   513   
3  2985  585  1337   0.438   179   486    0.368  674  792   0.851    62   317   
4  2482  669  1489   0.449   157   414    0.379  425  512   0.830   134   326   

   REB  AST  STL  BLK  TOV   PF   PTS  AST_TOV  STL_TOV  season_start_year  
0  640  374  116  105  280  143  2280     1.34     0.41               2012  
1  433  469  106   25  287  173  2133     1.63     0.37               2012  
2  610  551  129   67  226  110  2036     2.44     0.57               2012  
3  379  455  142   38  295  178  2023     1.54     0.48               2012  
4  460  171   52   32  175  205  1920     0.98     0.30               2012

As you can see the columns **season_start_year**, was sucessfully created

Note: The **New Orleans Hornets (NOH)** were sold to Tom Benson on **April 13, 2012.**

After purchasing the team, Benson indicated that he wanted **to change the club's nickname** to something more regionally appropriate.

On **December 14, 2012**, it was reported that the Hornets would change their nickname to the **New Orleans Pelicans.**

In [19]:
# Checking the array of the NBA teams by applying the .unique() method,

nba_data.TEAM.unique()

array(['OKC', 'LAL', 'MIA', 'HOU', 'NYK', 'GSW', 'MIL', 'POR', 'TOR',
       'BKN', 'CHA', 'LAC', 'BOS', 'UTA', 'PHI', 'IND', 'SAS', 'ATL',
       'CLE', 'NOH', 'DET', 'CHI', 'SAC', 'DAL', 'DEN', 'MEM', 'PHX',
       'ORL', 'MIN', 'WAS', 'NOP'], dtype=object)

**New Orleans Hornets = (NOH)**

**New Orleans Pelicans = (NOP)**

In [20]:
# Checking the total array of the NBA teams by applying the nunique() method,

nba_data.TEAM.nunique()

31

Create a new standar franchise name for the team New Orleans (NO)

In [21]:
# Standardize the NBA teams.

# Create a new standar franchise name for the team New Orleans

# New Orleans Hornets = (NOH)

# New Orleans Pelicans = (NOP)

nba_data['TEAM'].replace(to_replace=['NOP','NOH'], value='NO', inplace=True)

In [22]:
# Checking the array of the NBA teams by applying the .unique() method,
# with the new New Orleans (NO)

nba_data.TEAM.unique()

array(['OKC', 'LAL', 'MIA', 'HOU', 'NYK', 'GSW', 'MIL', 'POR', 'TOR',
       'BKN', 'CHA', 'LAC', 'BOS', 'UTA', 'PHI', 'IND', 'SAS', 'ATL',
       'CLE', 'NO', 'DET', 'CHI', 'SAC', 'DAL', 'DEN', 'MEM', 'PHX',
       'ORL', 'MIN', 'WAS'], dtype=object)

In [23]:
# Checking the total array of the NBA teams by applying the nunique() method,

nba_data.TEAM.nunique()

30

After standardize the NBA teams.

We can see that now we have **30 teams**

**New Orleans Hornets = (NO)**

In [24]:
# Let's clean and replace the Regular%20Season string that I found on the NBA API url
# Regular Season = RS

nba_data['Season_type'].replace('Regular%20Season','RS', inplace=True)

In [25]:
# Print the first 5 rows of the NBA dataframe by applying the .head() method,
# This method is will display the top 5 observations of the dataset

nba_data.head()

Year Season_type  PLAYER_ID           PLAYER     TEAM_ID TEAM  GP   MIN  \
0  2012-13          RS     201142     Kevin Durant  1610612760  OKC  81  3119   
1  2012-13          RS        977      Kobe Bryant  1610612747  LAL  78  3013   
2  2012-13          RS       2544     LeBron James  1610612748  MIA  76  2877   
3  2012-13          RS     201935     James Harden  1610612745  HOU  78  2985   
4  2012-13          RS       2546  Carmelo Anthony  1610612752  NYK  67  2482   

   FGM   FGA  FG_PCT  FG3M  FG3A  FG3_PCT  FTM  FTA  FT_PCT  OREB  DREB  REB  \
0  731  1433   0.510   139   334    0.416  679  750   0.905    46   594  640   
1  738  1595   0.463   132   407    0.324  525  626   0.839    66   367  433   
2  765  1354   0.565   103   254    0.406  403  535   0.753    97   513  610   
3  585  1337   0.438   179   486    0.368  674  792   0.851    62   317  379   
4  669  1489   0.449   157   414    0.379  425  512   0.830   134   326  460   

   AST  STL  BLK  TOV   PF   PTS  AST_TOV  STL_TOV  season_start_year  
0  374  116  105  280  143  2280     1.34     0.41               2012  
1  469  106   25  287  173  2133     1.63     0.37               2012  
2  551  129   67  226  110  2036     2.44     0.57               2012  
3  455  142   38  295  178  2023     1.54     0.48               2012  
4  171   52   32  175  205  1920     0.98     0.30               2012

Now we can see that the Regular Season = RS

# Creating 2 more data frames:

# 1- Dataframe for Regular Season = regular_season

# 2- Dataframe for Playoffs = playoffs

In [26]:
# Creating Regular Seasons dataframe

regular_season_dataframe = nba_data[nba_data['Season_type']=='RS']


# playoffs_df = data[data['Season_type']=='Playoffs']

In [27]:
# Print the Regular Seasons tables dataframe

regular_season_dataframe

Year Season_type  PLAYER_ID           PLAYER     TEAM_ID TEAM  GP  \
0     2012-13          RS     201142     Kevin Durant  1610612760  OKC  81   
1     2012-13          RS        977      Kobe Bryant  1610612747  LAL  78   
2     2012-13          RS       2544     LeBron James  1610612748  MIA  76   
3     2012-13          RS     201935     James Harden  1610612745  HOU  78   
4     2012-13          RS       2546  Carmelo Anthony  1610612752  NYK  67   
...       ...         ...        ...              ...         ...  ...  ..   
7071  2021-22          RS    1630207      Nate Hinton  1610612754  IND   2   
7072  2021-22          RS    1626155       Sam Dekker  1610612761  TOR   1   
7073  2021-22          RS    1629309   Trayvon Palmer  1610612765  DET   1   
7074  2021-22          RS    1629788       Tyler Hall  1610612752  NYK   1   
7075  2021-22          RS    1629597   Zylan Cheatham  1610612762  UTA   1   

       MIN  FGM   FGA  FG_PCT  FG3M  FG3A  FG3_PCT  FTM  FTA  FT_PCT  OREB  \
0     3119  731  1433   0.510   139   334    0.416  679  750   0.905    46   
1     3013  738  1595   0.463   132   407    0.324  525  626   0.839    66   
2     2877  765  1354   0.565   103   254    0.406  403  535   0.753    97   
3     2985  585  1337   0.438   179   486    0.368  674  792   0.851    62   
4     2482  669  1489   0.449   157   414    0.379  425  512   0.830   134   
...    ...  ...   ...     ...   ...   ...      ...  ...  ...     ...   ...   
7071     2    0     1   0.000     0     0    0.000    0    0   0.000     0   
7072     1    0     0   0.000     0     0    0.000    0    0   0.000     0   
7073    17    0     1   0.000     0     0    0.000    0    0   0.000     0   
7074     2    0     0   0.000     0     0    0.000    0    0   0.000     0   
7075     5    0     3   0.000     0     2    0.000    0    0   0.000     0   

      DREB  REB  AST  STL  BLK  TOV   PF   PTS  AST_TOV  STL_TOV  \
0      594  640  374  116  105  280  143  2280     1.34     0.41   
1      367  433  469  106   25  287  173  2133     1.63     0.37   
2      513  610  551  129   67  226  110  2036     2.44     0.57   
3      317  379  455  142   38  295  178  2023     1.54     0.48   
4      326  460  171   52   32  175  205  1920     0.98     0.30   
...    ...  ...  ...  ...  ...  ...  ...   ...      ...      ...   
7071     0    0    0    0    0    1    0     0     0.00     0.00   
7072     0    0    0    0    0    0    0     0     0.00     0.00   
7073     2    2    0    0    0    1    2     0     0.00     0.00   
7074     0    0    0    0    0    0    0     0     0.00     0.00   
7075     0    0    0    0    0    0    0     0     0.00     0.00   

      season_start_year  
0                  2012  
1                  2012  
2                  2012  
3                  2012  
4                  2012  
...                 ...  
7071               2021  
7072               2021  
7073               2021  
7074               2021  
7075               2021  

[5148 rows x 29 columns]

As we can see that the Regular Season dataframe contains 5148 rows × 29 columns

In [28]:
# Creating Playoffs dataframe

playoffs_dataframe = nba_data[nba_data['Season_type']=='Playoffs']

In [29]:
# Print the Playoffs dataframe tables dataframe

playoffs_dataframe

Year Season_type  PLAYER_ID           PLAYER     TEAM_ID TEAM  GP  \
468   2012-13    Playoffs       2544     LeBron James  1610612748  MIA  23   
469   2012-13    Playoffs       2225      Tony Parker  1610612759  SAS  21   
470   2012-13    Playoffs       1495       Tim Duncan  1610612759  SAS  21   
471   2012-13    Playoffs     202331      Paul George  1610612754  IND  19   
472   2012-13    Playoffs       2548      Dwyane Wade  1610612748  MIA  22   
...       ...         ...        ...              ...         ...  ...  ..   
7288  2021-22    Playoffs    1629006      Josh Okogie  1610612750  MIN   1   
7289  2021-22    Playoffs    1630556  Kessler Edwards  1610612751  BKN   2   
7290  2021-22    Playoffs    1630201    Malachi Flynn  1610612761  TOR   6   
7291  2021-22    Playoffs     202693  Markieff Morris  1610612748  MIA   2   
7292  2021-22    Playoffs     200794     Paul Millsap  1610612755  PHI   1   

      MIN  FGM  FGA  FG_PCT  FG3M  FG3A  FG3_PCT  FTM  FTA  FT_PCT  OREB  \
468   960  212  432   0.491    36    96    0.375  136  175   0.777    37   
469   765  167  365   0.458    11    31    0.355   87  112   0.777    14   
470   735  151  321   0.470     0     1    0.000   79   98   0.806    54   
471   780  119  277   0.430    34   104    0.327   93  128   0.727    16   
472   782  144  315   0.457     1     4    0.250   60   80   0.750    38   
...   ...  ...  ...     ...   ...   ...      ...  ...  ...     ...   ...   
7288    2    0    0   0.000     0     0    0.000    0    0   0.000     0   
7289    7    0    0   0.000     0     0    0.000    0    0   0.000     0   
7290   36    0    7   0.000     0     3    0.000    0    0   0.000     2   
7291    3    0    1   0.000     0     0    0.000    0    0   0.000     0   
7292    6    0    0   0.000     0     0    0.000    0    0   0.000     0   

      DREB  REB  AST  STL  BLK  TOV  PF  PTS  AST_TOV  STL_TOV  \
468    156  193  152   41   18   70  43  596     2.17     0.59   
469     54   68  146   24    3   53  26  432     2.75     0.45   
470    160  214   40   18   34   42  53  381     0.95     0.43   
471    125  141   96   25    9   75  72  365     1.28     0.33   
472     64  102  105   38   23   58  45  349     1.81     0.66   
...    ...  ...  ...  ...  ...  ...  ..  ...      ...      ...   
7288     0    0    0    0    0    0   0    0     0.00     0.00   
7289     0    0    1    1    0    1   3    0     1.00     1.00   
7290     1    3    3    1    0    1   6    0     3.00     1.00   
7291     1    1    0    0    0    1   2    0     0.00     0.00   
7292     1    1    1    0    0    0   1    0     0.00     0.00   

      season_start_year  
468                2012  
469                2012  
470                2012  
471                2012  
472                2012  
...                 ...  
7288               2021  
7289               2021  
7290               2021  
7291               2021  
7292               2021  

[2145 rows x 29 columns]

As we can see that the Regular Season dataframe contains 2145 rows × 29 columns

# NBA Glossary

In order to understand better the NBA dataset columns, here is the Glossary

https://www.nba.com/stats/help/glossary

In [30]:
# Creating NBA data col

nba_data.columns

Index(['Year', 'Season_type', 'PLAYER_ID', 'PLAYER', 'TEAM_ID', 'TEAM', 'GP',
       'MIN', 'FGM', 'FGA', 'FG_PCT', 'FG3M', 'FG3A', 'FG3_PCT', 'FTM', 'FTA',
       'FT_PCT', 'OREB', 'DREB', 'REB', 'AST', 'STL', 'BLK', 'TOV', 'PF',
       'PTS', 'AST_TOV', 'STL_TOV', 'season_start_year'],
      dtype='object')

There are columns that are totals, ratios, or percentages.

Example:

The column called **Field Goals Attempted (FGA)** is a total.

The column called **Field Goal Percentage (FG_PCT)** is the result of two attempts.

I want to create a list of the correct columns to use and conduct a group by.

In [31]:
total_columns = ['MIN','FGM','FGA','FG3M','FG3A','FTM','FTA',
              'OREB','DREB','REB','AST','STL','BLK','TOV','PF','PTS']

**Glossary:**

**MIN** = Minutes Played. The number of minutes played by a player or team.

**FGM** = Field Goals Made. The number of field goals that a player or team has made. This includes both 2 pointers and 3 pointers.

**FGA** = Field Goals Attempted. The number of field goals that a player or team has attempted. This includes both 2 pointers and 3 pointers.

**FG3M** = Three-point field goals made.

**FG3A** = Three-point field goals attempted.

**FTM** = Free Throws Made. The number of free throws that a player or team has made.

**FTA** = Free Throws Attempted. The number of free throws that a player or team has attempted.

**OREB** = Offensive Rebounds. The number of rebounds a player or team has collected while they were on offense.

**DREB** = Defensive Rebound. The number of rebounds a player or team has collected while they were on defensive.

**REB** = Rebounds. A rebound occurs when a player recovers the ball after a missed shot. This statistic is the number of total rebounds a player or team has collected on either offense or defense.

**AST** = Assists. The number of assists -- passes that lead directly to a made basket -- by a player.

**STL** = Steals. Number of times a defensive player or team takes the ball from a player on offense, causing a turnover.

**BLK** = Blocks. A block occurs when an offensive player attempts a shot, and the defense player tips the ball, blocking their chance to score.

**TOV** = Turnovers. A turnover occurs when the player or team on offense loses the ball to the defense.

**PF** = Personal Fouls. The number of personal fouls a player or team committed.

**PTS** = Points. The number of points scored.

# Data Analysis

## Which player stats are correlated with each other?